solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java

```diff
@@ -299,10 +299,19 @@ private PreparedRequest preparePutOrPost(
       InputStream is = streams.iterator().next().getStream();
       bodyPublisher = HttpRequest.BodyPublishers.ofInputStream(() -> is);
     } else if (queryParams != null && urlParamNames != null) {
-      ModifiableSolrParams requestParams = queryParams;
-      queryParams = calculateQueryParams(urlParamNames, requestParams);
-      queryParams.add(calculateQueryParams(solrRequest.getQueryParams(), requestParams));
-      bodyPublisher = HttpRequest.BodyPublishers.ofString(requestParams.toString());
+      // move params specified in urlParamNames from queryParams into urlParams
+      ModifiableSolrParams urlParams = calculateQueryParams(urlParamNames, queryParams);
+
+      // note this is only triggered if urlParamNames is set too - this this correct?
+      urlParams.add(calculateQueryParams(solrRequest.getQueryParams(), queryParams));
```

Comment:
I'm unclear when this would kick in - it can't trigger if `urlParamNames` is null.

Reply:
Indeed `urlParamNames` is never null (see `HttpSolrClientBase` constructor line 107ff), so the null check is spurious.  Should you comment-out line 306, you will see `testQueryString` fail in the case where there are query parameters but nothing was set for `urlParamNames` (using the `withTheseParamNamesInTheUrl` option).

Reply:
`queryParams` can't be null either, can it? I've made this the `else` block - I added an `UnsupportedOperationException` in the original `noBody()` one and saw it was never triggered by tests. Does this change look good to you?

Reply:
Well, you are right, it does not seem urlParams or queryParams can ever be null.  So you're correct, the "noBody" case never gets touched.  I guess I had thought there would be a case where we would could post with an empty body but clearly this cannot happen.